sw/host/opentitanlib/src/proxy/async_uart.rs

```diff
@@ -0,0 +1,149 @@
+        &mut self,
+        event: &Event,
+        connection_map: &mut HashMap<Token, Connection>,
+    ) -> Result<bool> {
+        match self.token_map.get(&event.token()) {
+            Some(uart_key) => {
+                if event.is_readable() {
+                    self.async_uarts
+                        .get_mut(uart_key)
+                        .unwrap()
+                        .process(connection_map)?;
+                }
+                Ok(true)
+            }
+            None => Ok(false),
+        }
+    }
+}
+
+pub struct AsyncUart {
+    conn_tokens: HashSet<Token>,
+    uart: std::rc::Rc<dyn crate::io::uart::Uart>,
+    channel: u32,
+}
+
+impl AsyncUart {
+    pub fn new(uart: &std::rc::Rc<dyn crate::io::uart::Uart>, channel: u32) -> Self {
+        Self {
+            conn_tokens: HashSet::new(),
+            uart: uart.clone(),
+            channel,
+        }
+    }
+
+    pub fn add_connection(&mut self, conn_token: Token) {
+        self.conn_tokens.insert(conn_token);
+    }
+
+    fn process(&mut self, connection_map: &mut HashMap<Token, Connection>) -> Result<()> {
+        let mut buf = [0u8; 256];
+        // Timeout is just a safety, as we should be here only if the uart has data ready to read.
+        let len = self
+            .uart
+            .read_timeout(&mut buf, Duration::from_millis(10))?;
+        // Iterate through list of connections who "subscribe" to this UART instance, putting data
+        // into each of their outgoing buffers.  If any connection has been closed, remove the
+        // reference to it from `conn_tokens`.
+        let mut closed_connections = Vec::new();
+        for conn_token in &self.conn_tokens {
+            if let Some(conn) = connection_map.get_mut(conn_token) {
+                conn.transmit_outgoing_msg(super::protocol::Message::Async {
+                    channel: self.channel,
+                    msg: super::protocol::AsyncMessage::UartData {
+                        data: buf[0..len].to_vec(),
+                    },
+                })?
+            } else {
+                // The particular TCP connection has been closed, and should be removed from the
+                // list of subsribers to this `Uart`.  Keep its ID for removal when this
+                // iteration is done.
+                closed_connections.push(*conn_token);
+            }
+        }
+        for conn_token in closed_connections {
+            self.conn_tokens.remove(&conn_token);
+        }
+        Ok(())
+    }
+}
+
+/// Struct allowing `HashMap` to use `Uart` object identity as the key.
+#[derive(Clone)]
+pub struct UartKey(pub Rc<dyn Uart>);
+
+impl Hash for UartKey {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        hash(&*self.0, state)
+    }
+}
+impl PartialEq for UartKey {
+    /// Determines whether the two `Rc`s point to the same `Uart` instance.
+    fn eq(&self, other: &Self) -> bool {
+        &*self.0 as *const dyn Uart == &*other.0 as *const dyn Uart
```

Comment:
I think you could use the `as_ptr` function on `Rc<T>`:
```
self.0.as_ptr() == other.0.as_ptr()
```

Reply:
It has been a while, but I remember getting CI errors on all sort of variations, something about dyn vtable pointers not being guaranteed to be unique if multiple compilation units could be independently converting concrete types into `dyn Uart`.  (This will be unlikely to be the case, since each individual transport converts its own types into the trait only in its own file.)
I do not recall how I arrived at `&*self.0 as *const dyn Uart == &*other.0 as *const dyn Uart`, and I do not believe that it solves the problem that the CI tried to warn me against, at most obscure the operation enough that the CI does not recognize it.
Let me try if your suggestion goes through CI.